### PR TITLE
[v2] Added Azure Log Analytics scaler

### PR DESCRIPTION
**Pull request by spoplavskiy**
### Checklist

- [X] Commits are signed with Developer Certificate of Origin (DCO)
- [X] Tests have been added
- [X] A [PR 244](https://github.com/kedacore/keda-docs/pull/244) is opened to update the documentation on https://github.com/kedacore/keda-docs
- [x] Changelog has been updated

This Scaler related to Feature Request [1061](https://github.com/kedacore/keda/issues/1061)
It implements all features and use-cases described there.

Signed-off-by: Sergiy Poplavskyi <spopla@microsoft.com>

### CODE REVIEW

**Comment by tomkerkhove** [comment]
Can you update https://github.com/kedacore/keda/blob/v2/CHANGELOG.md as well please?

**Comment by spoplavskiy** [comment]
> Can you update https://github.com/kedacore/keda/blob/v2/CHANGELOG.md as well please?

@tomkerkhove , thanks for pointing, added in last commit

**Review by zroubalik — COMMENTED** [review]
@spoplavskiy thanks for the contribution! I don't feel confident enough to review Azure related scalers, will keep it for the others. But I wonder whether it is doable to add e2e test for this type of scaler?

https://github.com/kedacore/keda/tree/v2/tests

**Comment by spoplavskiy** [comment]
> @spoplavskiy thanks for the contribution! I don't feel confident enough to review Azure related scalers, will keep it for the others. But I wonder whether it is doable to add e2e test for this type of scaler?
> 
> https://github.com/kedacore/keda/tree/v2/tests

@zroubalik , yes, I can try. The thing is, that Log Analytics workspace should be created on Azure subscription, used by your build pipeline. Also, a Service Principal should have access to this workspace. I see in ".env", that there is variables, related to Service Principal. Can I introduce "TEST_LOG_ANALYTICS_WORKSPACE_ID" variable? I will commit empty value (same, as others), but it should be created and added in Build Pipeline.

**Comment by zroubalik** [comment]
Great! I think that @ahmelsayed or @jeffhollan could help you with that

**Comment by spoplavskiy** [comment]
> Great! I think that @ahmelsayed or @jeffhollan could help you with that

@zroubalik , I have added E2E tests from Log Analytics scaler. Also, added support for 'FromEnv' definitions. I had a conversation with @ahmelsayed, provided him all required scripts to create LA workspace and grant access to Service Principal on subscription, used for tests.

**Comment by tomkerkhove** [comment]
> > Great! I think that @ahmelsayed or @jeffhollan could help you with that
> 
> @zroubalik , I have added E2E tests from Log Analytics scaler. Also, added support for 'FromEnv' definitions. I had a conversation with @ahmelsayed, provided him all required scripts to create LA workspace and grant access to Service Principal on subscription, used for tests.

Thanks @spoplavskiy!

**Comment by zroubalik** [comment]
@spoplavskiy this is great, e2e tests are really important, thanks! I'll keep the final review on @ahmelsayed 

Just a minor nit, could you please rebase the PR, there's a conflict in the Changelog.

**Comment by spoplavskiy** [comment]
> @spoplavskiy this is great, e2e tests are really important, thanks! I'll keep the final review on @ahmelsayed
> 
> Just a minor nit, could you please rebase the PR, there's a conflict in the Changelog.

@zroubalik , done. But now it looks like I break static check. For some reason "pre-commit" tool thinks, that I have changed test file fully, and fails on check. Is that OK in this case?

File: pkg/scalers/azure_log_analytics_scaler.go [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+		tokenInfo, err := s.refreshAccessToken()
+		if err != nil {
+			return tokenData{}, err
+		}
+
+		s.cache.accessToken = tokenInfo
+		return tokenInfo, nil
+	}
+
+	logAnalyticsLog.Info(fmt.Sprintf("Reuse existing access token..."))
```
Comment (zroubalik) [inline pkg/scalers/azure_log_analytics_scaler.go:265]: No need to call fmt.Sprintf() inside.

File: pkg/scalers/azure_log_analytics_scaler.go [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+
+	body, statusCode, err := s.executeLogAnalyticsREST(query, tokenInfo)
+
+	if body == nil || len(body) == 0 {
+		return metricsData{}, fmt.Errorf("Error executing Log Analytics REST API request: empty body")
+	}
+
+	//Handle expired token
+	if statusCode == 403 || strings.Contains(string(body), "TokenExpired") {
+		logAnalyticsLog.Info(fmt.Sprintf("Token expired, refreshing token..."))
```
Comment (zroubalik) [inline pkg/scalers/azure_log_analytics_scaler.go:280]: No need to call `fmt.Sprintf()`

File: pkg/scalers/azure_log_analytics_scaler.go [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+	if err != nil {
+		return metricsData{}, err
+	}
+
+	metricsInfo, err := s.executeQuery(s.metadata.query, tokenInfo)
+	if err != nil {
+		return metricsData{}, err
+	}
+
+	logAnalyticsLog.Info(fmt.Sprintf("Providing metric value: %d", metricsInfo.value))
```
Comment (zroubalik) [inline pkg/scalers/azure_log_analytics_scaler.go:248]: We are using [Structured logging](https://www.client9.com/structured-logging-in-golang/).

Would be better to replace the call to something like: 
logAnalyticsLog.Info("Getting metric value, "metrics value", metricsInfo.value))

File: pkg/scalers/azure_log_analytics_scaler.go [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+	if err != nil {
+		return tokenData{}, err
+	}
+
+	//Now, let's check we can use this token. If no, wait until we can use it
+	currentTimeSec := time.Now().Unix()
+	if currentTimeSec < tokenInfo.NotBefore {
+		if currentTimeSec < tokenInfo.NotBefore+10 {
+			sleepDurationSec := int(tokenInfo.NotBefore - currentTimeSec + 1)
+			logAnalyticsLog.Info(fmt.Sprintf("AAD token not ready, waiting %d seconds before continue...", sleepDurationSec))
```
Comment (zroubalik) [inline pkg/scalers/azure_log_analytics_scaler.go:385]: Same as the others log messages

File: pkg/scalers/azure_log_analytics_scaler.go (hunk 3 above)
Comment (zroubalik) [reply pkg/scalers/azure_log_analytics_scaler.go:248]: If this is DEBUG log message, would be better to call it `logAnalyticsLog.V(1).Info("Getting metr...)`
So we don't flood output with this particular log

**Review by zroubalik — CHANGES_REQUESTED** [review]
Minor nits on logging

File: pkg/scalers/azure_log_analytics_scaler.go (hunk 1 above)
Comment (spoplavskiy) [reply pkg/scalers/azure_log_analytics_scaler.go]: Thanks for pointing. I have removed this entry in last commit, as it is not so valuable in this case.

File: pkg/scalers/azure_log_analytics_scaler.go (hunk 2 above)
Comment (spoplavskiy) [reply pkg/scalers/azure_log_analytics_scaler.go]: Fixed in last commit

File: pkg/scalers/azure_log_analytics_scaler.go (hunk 3 above)
Comment (spoplavskiy) [reply pkg/scalers/azure_log_analytics_scaler.go]: Thanks for pointing. I have changed this definition in last commit.

File: pkg/scalers/azure_log_analytics_scaler.go (hunk 4 above)
Comment (spoplavskiy) [reply pkg/scalers/azure_log_analytics_scaler.go]: Fixed in last commit

**Comment by spoplavskiy** [comment]
> Minor nits on logging

@zroubalik , thanks for pointing. All your suggestions has been implemented in last commit.

File: pkg/scalers/azure_log_analytics_scaler.go [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+			logAnalyticsLog.Error(err, "Error processing Log Analytics query")
+			return err
+		}
+
+		s.cache.metricValue = receivedMetric.value
+
+		if receivedMetric.threshold > 0 {
+			s.cache.metricThreshold = receivedMetric.threshold
+		} else {
+			s.cache.metricThreshold = s.metadata.threshold
```
Comment (ahmelsayed) [inline pkg/scalers/azure_log_analytics_scaler.go:230]: what does `receivedMetric.threshold` represent here? All other scalers use the threshold value defined in the metadata for updating the HPA config. I'm trying to understand if supplying the threshold value from Azure Log Analytics makes sense or not here.

File: pkg/scalers/azure_log_analytics_scaler.go [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+
+	err = json.NewDecoder(bytes.NewReader(body)).Decode(&queryData)
+	if err != nil {
+		if statusCode != 200 {
+			return metricsData{}, fmt.Errorf("Error processing Log Analytics REST API request: %d", statusCode)
+		}
+		return metricsData{}, fmt.Errorf("Can't decode JSON from Log Analytics REST API result: %v", err)
+	}
+
+	if statusCode == 200 {
```
Comment (ahmelsayed) [inline pkg/scalers/azure_log_analytics_scaler.go:301]: nit: can this be changed into an early return. i.e: 

```
if statusCode != 200 {
    return .....
}
```

File: pkg/scalers/azure_log_analytics_scaler.go [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+		//Pre-validation of query result:
+		if len(queryData.Tables) == 0 || len(queryData.Tables[0].Columns) == 0 || len(queryData.Tables[0].Rows) == 0 {
+			return metricsData{}, fmt.Errorf("There is no results after running your query")
+		} else if len(queryData.Tables) > 1 {
+			return metricsData{}, fmt.Errorf("Too many tables in query result: %d. Expected: 1", len(queryData.Tables))
+		} else if len(queryData.Tables[0].Rows) > 1 {
+			return metricsData{}, fmt.Errorf("Too many rows in query result: %d. Expected: 1", len(queryData.Tables[0].Rows))
+		}
+
+		if len(queryData.Tables[0].Rows) == 1 {
```
Comment (ahmelsayed) [inline pkg/scalers/azure_log_analytics_scaler.go:316]: wouldn't this always be true based on the pre-validation above?

**Comment by ahmelsayed** [comment]
I added the required secret to the pipeline. Just have a couple of minor comments on the PR.

File: pkg/scalers/azure_log_analytics_scaler.go (hunk 5 above)
Comment (spoplavskiy) [reply pkg/scalers/azure_log_analytics_scaler.go:230]: There is a possibility to set threshold, based on query result. If query returns row with 2 cells, 2d cell will be interpreted as threshold. Query result has higher priority, than threshold, set in metadata. With this, threshold in metadata can be overwritten. I described this logic and limitations in a documentation for scaler.

File: pkg/scalers/azure_log_analytics_scaler.go (hunk 6 above)
Comment (spoplavskiy) [reply pkg/scalers/azure_log_analytics_scaler.go:302]: Thanks for catching. Moved it upper in last commit.

File: pkg/scalers/azure_log_analytics_scaler.go (hunk 7 above)
Comment (spoplavskiy) [reply pkg/scalers/azure_log_analytics_scaler.go]: Yes, you are right, changed it in last commit. Thanks for catching.

**Comment by spoplavskiy** [comment]
> I added the required secret to the pipeline. Just have a couple of minor comments on the PR.

@ahmelsayed , thanks!

File: pkg/scalers/azure_log_analytics_scaler.go (hunk 5 above)
Comment (ahmelsayed) [reply pkg/scalers/azure_log_analytics_scaler.go:230]: That makes sense. I just took a look at the docs PR. Thanks for the detailed explanation there.

**Review by ahmelsayed — APPROVED** [review]

**Pull request merged**